WIP: Documentation automation (refs #2860)

ilyavolodin opened the pull request:
This is a WIP pull request to check schema of documentation object. See #2860 for more details.

nzakas commented on lib/rules/accessor-pairs.js:
```diff
@@ -63,3 +63,11 @@ module.exports = function(context) {
 };
+
+module.exports.documentation = {
+    errors: ["Getter is not present", "Setter is not present"],
```
I don't think it's practical or wise to require error messages to be out here. There's a significant chance of forgetting to keep these in sync with actual code.

ilyavolodin replied:
Agree. What I was thinking is exposing this to the rule somehow, so that error messages can be recorded in meta area (better name), and then used inside the rule.

btmills replied:
> exposing this to the rule somehow

Something like this?

```js
module.exports.meta = {
    errors: {
        GETTER: "Getter is not present",
        SETTER: "Setter is not present"
    }
    // ...
};
```

From there, two options:

```js
// First option: core passes `errors` to the rule, which references the messages by key
context.report(node, context.errors.GETTER);
// Second opiton: rule references the error message in `context.report` by key
context.report(node, "GETTER");
```

The first looks cleaner to me, but the second would do a better job of enforcing that rules don't supply their own error messages out of sync with the metadata.

ilyavolodin replied:
Second one would probably be a bit better, but it would break all of the plugins. I would rather go for backwards compatible one. So first one in this case.

btmills replied:
Agreed. If it really becomes a problem, we could add a test step that checks errors generated by the tests against the messages supplied in `meta.errors`.

ilyavolodin replied:
I actually think that since we added schema to rules, we should create an eslint-plugin just for rules and consume it. We could check for correct schema and documentation with it.

btmills replied:
Interesting... how much would that offload from eslint-tester?

ilyavolodin replied:
Probably not a whole lot. It would be in addition to eslint-tester. Although, if we were to do something like what I suggested below, about parsing options through schema, that will allow us to validate options through the custom rule, instead of eslint-tester.

nzakas replied:
I still think having errors in this section is a bad idea. Our output already shows the rule the message comes from, I don't think we really need to force people into a situation like this.

nzakas commented on lib/rules/accessor-pairs.js:
```diff
@@ -63,3 +63,11 @@ module.exports = function(context) {
+    description: "enforces getter/setter pairs in objects",
+    ruleType: "Best Practices",
+    recommended: false,
+    default: [{getWithoutSet: false, setWithoutGet: true}]
```
Also a bit concerned about this, unless we automatically pass the defaults directly to the rule, this will become out of sync with code.

ilyavolodin replied:
The other way is to require this as part of the schema. I really think it would help a lot of users to clearly expose defaults for all rules. Right now, not all rules document this in a easy to find way.

ilyavolodin replied:
@btmills Any ideas on how we might be able to mark defaults in the `schema`? Is it possible at all?

btmills replied:
There's an official [`default`](http://json-schema.org/latest/json-schema-validation.html#anchor101) metadata property in the spec. Since it's freeform, we'd want to add a step in the tests to parse the supplied value and verify it against the schema.

btmills replied:
That alone doesn't solve the synchronization issue since rules sometimes do their own processing to merge options with defaults. Would using [merge](https://www.npmjs.com/package/merge) with `options = merge.recursive(true, schema.default, context.options)` do the trick?

ilyavolodin replied:
Hmmm.. What if we were to change the way options are accessible in the rules. Schema would parse an array and create an object with all options available, then inside the rule, instead of doing ```var getWithoutSet = context.options[0]``` you would be able to do ```context.options.getWithoutSet```?

nzakas replied:
That's what I meant, if you're going to force people to specify default options, then we need to handle passing those options from the core.

nzakas commented on lib/rules/accessor-pairs.js:
```diff
@@ -63,3 +63,11 @@ module.exports = function(context) {
 
 };
+
+module.exports.documentation = {
```
I'd call this "meta" as some of this stuff is more than just docs related.

ilyavolodin replied:
Agree

btmills replied:
I like `meta`. Should `schema` be moved under `meta.schema`? (I haven't thought about a backwards-compatible way to do that yet, so it may not be possible.)

ilyavolodin replied:
`schema` is not really `meta` it actually serves a purpose inside the rule itself. So I would leave it as is for now.

ilyavolodin commented:
I updated pull request with proposed changes for schema and default values. Obviously build fails because this change breaks every rule that has options, but I updated 4 rules that I originally modified. I'm not sure rule-context is the right place to do this in, but looking for feedback.

nzakas commented:
Implementation looks good. I'm still having a hard time believing it makes sense to put error messages in this spot. What is the value of that given that every error message is already output with the rule name?

ilyavolodin commented:
Our default reporter does output rule names, however, that doesn't mean all others will do the same. In fact, most of the IDE integrations that I've seen for ESLint do not report rule name by default. I'll re-arrange error messages in such a way that `meta` can be used as a store for messages, this way rules can just retrieve error messages from `meta` instead of duplicating the string.

nzakas commented:
I'd like to see a full example of what a rule should look like with the schema, meta, and rule info. I'm worried that we are piling on more requirements to rule authors and I'd like to see what we are now asking of them in totality. Maybe take one simple rule and convert it completely?

There's also the problem of converting every error we have, as that looks to be a completely manual process.

ilyavolodin commented:
Fair enough. I pushed a change for `array-bracket-spacing` that should be fully updated for new syntax.

nzakas commented on lib/rules/array-bracket-spacing.js:
```diff
@@ -130,19 +85,19 @@ module.exports = function(context) {
         if (isSameLine(first, second)) {
             if (openingBracketMustBeSpaced && !isSpaced(first, second)) {
-                reportRequiredBeginningSpace(node, first);
+                reportError(node, context.errors.spaceBefore, first);
```
I don't think `context` is should be doing this. It's job is to provide information that isn't otherwise available in the file. This info exists in the file already.

ilyavolodin replied:
I agree with the above statement, but if we want to support i18n, then error messages should probably be passed from context. That would allow us to override messages from the core.

nzakas replied:
I'm not sure there's a straight line between translations and changing what context is meant to do. I'd need to see a more fully fleshed out proposal for i18n.

nzakas commented on lib/rules/array-bracket-spacing.js:
```diff
@@ -160,21 +115,37 @@ module.exports = function(context) {
+
+module.exports.meta = {
+    errors: {
+        extraSpaceAfter: "There should be no space after '{{token}}'",
```
I have a hard time believing we'll get people to define their error messages at the very bottom of every file. This is part of the developer ergonomics - it's counterintuitive. Most people define their messages either at the top of the file or right inline with `context.report()`. This represents too dramatic a shift, IMHO. We need to have a better recommendation.

ilyavolodin replied:
Ok, that was just one of the approaches. I'll try to think of some other way, and submit another pull request.
